Codec/Archive/Tar/Check/Internal.hs

```diff
@@ -65,12 +66,17 @@ checkSecurity = checkEntries checkEntrySecurity
 
 -- | @since 0.6.0.0
 checkEntrySecurity :: Entry -> Maybe FileNameError
-checkEntrySecurity entry = case entryContent entry of
-    HardLink     link -> check (entryPath entry)
-                 `mplus` check (fromLinkTarget link)
-    SymbolicLink link -> check (entryPath entry)
-                 `mplus` check (fromLinkTarget link)
-    _                 -> check (entryPath entry)
+checkEntrySecurity entry = check (entryPath entry) <|> case entryContent entry of
+    HardLink     link -> check (fromLinkTarget link)
+    SymbolicLink link -> check (FilePath.Native.takeDirectory (entryPath entry)
+                                FilePath.Native.</> fromLinkTarget link)
+    OtherEntryType 'K' longPath _
+                      -- We don't know yet whether it's a hard or a soft link.
+                      -- Let's err on the safe side and validate as a hard one.
+                      -> check (Char8.unpack $ Char8.takeWhile (/= '\0') longPath)
```

Comment:
Why are we not considering the basepath here? Will that not yield false-positives?

Reply:
See cases for hard and soft links above. AFAIU from the difference between `handleHardLinks` and `handleSymLinks` in `Unpack.hs`, hard links are "absolute" in the context of an archive (thus `check longPath`) and relative to the current position in the archive (thus `check $ takeDirectory tarPath </> longPath`). Here we don't know whether it's a hard or a soft link, so assuming the worst.

Reply:
What I mean is that symlink targets like `./../../bar` will throw here afaiu, because it's assumed to be the tar root.

However, it may be a safe symlink.

I've written a patch where we don't use `checkEntrySecurity` but only operate on `Entries e` and can look as far ahead as we want, similar to `unpack`. It's a bit involved, so I haven't finished yet.